File: src/data-scratch-library/dsl/c19_deep_learning/io.py
"""
Model serialization: save / load neural-network weights as JSON.
"""

import json
from typing import List

from .deep_learning import Tensor, shape
from .layer import Layer


def save_weights(model: Layer, filename: str) -> None:
    """Save model parameters to a JSON file."""
    weights = list(model.params())
    with open(filename, "w") as f:
        json.dump(weights, f)


def load_weights(model: Layer, filename: str) -> None:
    """Load model parameters from a JSON file (in-place)."""
    with open(filename) as f:
        weights = json.load(f)
    for param, weight in zip(model.params(), weights):
        if shape(param) != shape(weight):
            raise ValueError(
                f"Shape mismatch: expected {shape(param)}, got {shape(weight)}"
            )
    for param, weight in zip(model.params(), weights):
        param[:] = weight

File: src/data-scratch-node-library/ts-data-scratch/deep_learning/io.ts
// Model serialization utilities for deep learning framework
// Ported from Python io module

import { readFileSync, writeFileSync } from 'fs';
import { Layer } from './layer';
import { Tensor, tensorShape } from './index';

export function saveWeights(model: Layer, filename: string): void {
    const weights = model.params();
    const jsonData = JSON.stringify(weights, null, 2);
    writeFileSync(filename, jsonData, 'utf8');
}

export function loadWeights(model: Layer, filename: string): void {
    const jsonData = readFileSync(filename, 'utf8');
    const weights: Tensor[] = JSON.parse(jsonData);
    
    // Check for consistency
    const params = model.params();
    for (let i = 0; i < params.length; i++) {
        const param = params[i];
        const weight = weights[i];
        
        if (JSON.stringify(tensorShape(param)) !== JSON.stringify(tensorShape(weight))) {
            throw new Error(`Shape mismatch: expected ${JSON.stringify(tensorShape(param))}, got ${JSON.stringify(tensorShape(weight))}`);
        }
    }
    
    // Load weights using slice assignment equivalent
    for (let i = 0; i < params.length; i++) {
        copyTensor(weights[i], params[i]);
    }
}

function copyTensor(source: Tensor, target: Tensor): void {
    if (Array.isArray(source) && source.length > 0 && typeof source[0] === 'number') {
        // 1D tensor
        for (let i = 0; i < source.length; i++) {
            (target as number[])[i] = (source as number[])[i];
        }
    } else if (Array.isArray(source) && source.length > 0 && Array.isArray(source[0])) {
        // Multi-dimensional tensor
        for (let i = 0; i < source.length; i++) {
            copyTensor((source as Tensor[])[i], (target as Tensor[])[i]);
        }
    }
}
